kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncThread.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.api.async.internals;
+
+import dev.responsive.kafka.api.async.internals.contexts.AsyncThreadProcessorContext;
+import dev.responsive.kafka.api.async.internals.events.AsyncEvent;
+import dev.responsive.kafka.api.async.internals.queues.ProcessingQueue;
+import dev.responsive.kafka.api.async.internals.queues.ReadOnlyProcessingQueue;
+import java.io.Closeable;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.kafka.common.utils.LogContext;
+import org.slf4j.Logger;
+
+public class AsyncThread extends Thread implements Closeable {
+
+  private final Logger log;
+
+  private final AtomicBoolean shutdownRequested = new AtomicBoolean(false);
+
+  private final ReadOnlyProcessingQueue processingQueue;
+  private final Map<Integer, AsyncNodeContainer> nodeIdToContainer;
+
+  public AsyncThread(final String name, final ProcessingQueue processingQueue) {
+    super(name);
+    this.log = new LogContext(String.format("async-thread [%s] ", name)).logger(AsyncThread.class);
+
+    this.processingQueue = processingQueue;
+
+    // concurrent b/c new processors are added by StreamThread via AsyncThreadPool
+    this.nodeIdToContainer = new ConcurrentHashMap<>();
+  }
+
+  public void addProcessor(
+      final AsyncNodeContainer processorContainer
+  ) {
+    nodeIdToContainer.put(processorContainer.partition(), processorContainer);
+  }
+
+  public void removeProcessor(
+      final int partition
+  ) {
+    nodeIdToContainer.remove(partition);
+  }
+
+  @SuppressWarnings("unchecked")
+  public <KOut, VOut> AsyncThreadProcessorContext<KOut, VOut> context(final int partition) {
+    final var context = nodeIdToContainer.get(partition).asyncContext();
+    return (AsyncThreadProcessorContext<KOut, VOut>) context;
+  }
+
+  @Override
+  public void run() {
+    try {
+      while (!shutdownRequested.getOpaque()) {
+        processOneEvent();
+      }
+    } catch (final Exception e) {
```

Comment:
we should pass this exception back up if its not expected. The async processor should have the same semantics as the synchronous one when an error happens - the stream thread should just exit and run the unexpected exception handler.

Reply:
The only thing missing from the POC still is error handling. So yeah, I agree, I just didn't get around to it quite yet. I'm saving that for last basically